File: src/cradle.py
#!/usr/bin/env python3

import sys
import io

## Variable declarations
look = ''  # Lookahead character


## Function declarations


def get_char():
    """Read new character from input stream"""
    global look
    look = sys.stdin.read(1)


def report_error(x):
    """Report an error"""
    print(f"Error: {x}")


def report_abort(x):
    """Report error and halt"""
    report_error(x)
    sys.exit(1)


def report_expected(x):
    """Report what was expected"""
    report_abort(f"{x} Expected")


def match_char(x):
    """Match a specific input character"""
    if look == x:
        get_char()
    else:
        report_expected(f"' {x} '")


def is_alpha_char(x):
    """Recognize an alpha character"""
    return x.isalpha()


def is_digit_char(x):
    """Recognize a decimal digit"""
    return x.isdigit()


def is_addop_char(x):
    """Recognize an addop character"""
    return x in ['+', '-']


def get_identifier():
    """Get an identifier"""
    if not is_alpha_char(look):
        report_expected("Name")
    token = ''
    while is_alpha_char(look) or is_digit_char(look):
        token += look
        get_char()
    return token.to_upper()


def get_number():
    """Get a number"""
    if not is_digit_char(look):
        report_expected("Number")
    token = ''
    while is_digit_char(look):
        token += look
        get_char()
    return token


def emit(s):
    """Output a string with tab"""
    print(f"\t{x}")


def emit_ln(s):
    """Output a string with tab and CRLF"""
    print(f"\t{x}\n")


def initialize():
    """Initialize"""
    get_char()


def main():
    """Main program"""
    initialize()
    print("Hello, World!")


if __name__ == "__main__":
    main()

File: tutor1/cradle.c
/**
 * @file        cradle.c
 * @brief       Utility functions for the Cradle program
 * @date        November 13, 2025   
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "cradle.h"

/**
 * Get a character from the input stream
 */
void get_char(void)
{
    look = getchar();
}

/**
 * Report an error
 */
void report_error(const char *message)
{
    printf("ERROR: %s\n", message);
}

/**
 * Report an abort and exit the program
 */
void report_abort(const char *message)
{
    printf("ABORT: %s\n", message);
    exit(1);
}

/**
 * Report an expected character
 */
void report_expected(const char *message)
{
    printf("EXPECTED: %s\n", message);
    exit(1);
}

/**
 * Match a character
 */
void match_char(char c)
{
    if (look != c) {
        report_expected(c);
    }
    // Otherwise, get the next character
    get_char();
}

/**
 * Check if a character is an alpha character
 */
void is_alpha_char(char c)
{
    if (!isalpha(c)) {
        report_error("Alpha character expected");
    }
    printf("Alpha character matched: %c\n", c);
}

/**
 * Check if a character is a digit character
 */
void is_digit_char(char c)
{
    if (!isdigit(c)) {
        report_error("Digit character expected");
    }
    printf("Digit character matched: %c\n", c);
}

/**
 * Get a name
 */
char get_name(void)
{
    char c = look;
    if (!isalpha(c)) {
        report_expected("Name expected");
    }
    get_char();
    printf("Name matched: %c\n", c);
    return toupper(c);
}

/**
 * Get a number
 */
char get_num(void)
{
    char c = look;
    if (!isdigit(c)) {
        report_expected("Number expected");
    }
    get_char();
    printf("Number matched: %c\n", c);
    return c;
}

/**
 * Emit a string to the console
 */
void emit(const char *s)
{
    printf("\t%s", s);
}

/**
 * Emit a string to the console with a newline
 */
void emit_ln(const char *s)
{
    emit(s);
    printf("\n");
}

/**
 * Initialize the Cradle
 */
void initialize(void)
{
    get_char();
}

/**
 * Main function
 */
void main(void)
{
    initialize();
}